File: TnSeq/CalcDE.R
# Copyright 2013 Colm Ryan colm@colmryan.org
# License GPL v3 (http://www.gnu.org/licenses/gpl.txt) 


library('XLConnect')
library('edgeR')

#Control
controlName <- 'MH'

#Drugs
drugNames <- c('Van','Cip', 'DaptoCaCl2', 'DaptoNO', 'Gent', 'Lin', 'Ox', 'Rif', 'ST', 'Cipin', 'Gentin', 'Linin', 'Oxin', 'Rifin', 'STin') 
# drugNames <- c('Cipin', 'Gentin', 'Linin', 'Oxin', 'Rifin', 'STin'  )
# drugNames = c('Van', 'Cip')
DEOutFile = 'DEins.xlsx'

load_data <- function(fileName){
	# #Load the appropriate columns from the Excel sheet
	# tmpData <- readWorksheetFromFile(fileName, sheet="MainSheet", rownames=2)[, c("Start", "Length", "ReadCount")]

	#Strip the troublesome intergenic line from the file
	system(paste("sed -n '/^intergenic/ !p' < ", fileName, " > tmpFile", sep=''))
	tmpData <- read.table("tmpFile", sep="\t", header=TRUE, row.names=2, na.strings="", quote="")[, c("Start", "Length", "ReadCount")]
	#Drop the intergenic
	# tmpData <- tmpData[c(rownames(tmpData) != "intergenic"),]
	#Replace empty values by zeros
	tmpData[is.na(tmpData)] <- 0
	#Sort by start position for convenience
	tmpData <- tmpData[with(tmpData, order(Start)),] 
}

#Load some data
controlData <- list()
drugData <- list()
controlData$A <- load_data(paste('NEWAGG/', controlName, "A.rdp", sep=''))
controlData$B <- load_data(paste('NEWAGG/', controlName, "B.rdp", sep=''))
for (drugName in drugNames){
	print(drugName)
	drugData[[drugName]] <- list()
	drugData[[drugName]]$A <- load_data(paste('NEWAGG/', drugName, "A.rdp", sep=''))
	drugData[[drugName]]$B <- load_data(paste('NEWAGG/', drugName, "B.rdp", sep=''))
}


#Use the read density as a function of position to figure out position normalization
add_norm_counts <- function(geneDataIn, plotName){
	#Remove the bottom/top 10% to reduce influence of outliers
	tmpData <- subset(geneDataIn, ReadCount > quantile(geneDataIn$ReadCount, 0.1) & ReadCount < quantile(geneDataIn$ReadCount, 0.9))
	tmpData$readDensity <- tmpData$ReadCount/tmpData$Length
	# X11()
	png(filename=paste(plotName , '_Bowl.png'))
	plot(tmpData$Start, tmpData$readDensity)
	#Make a loess fit
	loessFit <- loess(readDensity~position, data.frame(position=tmpData$Start, readDensity=tmpData$readDensity), 
				control = loess.control(statistics = c("approximate"),trace.hat = c("approximate"), iterations=3))
	tmpData$smoothed <- predict(loessFit, tmpData$Start)
	lines(tmpData$Start, tmpData$smoothed, col="red", lwd=2)
	dev.off()
	geneDataIn$normRatio <- predict(loessFit, geneDataIn$Start)
	#Flat-line extrapolate past the end points if necessary
	firstNonNAN <- which.min(is.na(geneDataIn$normRatio))
	geneDataIn$normRatio[1:firstNonNAN-1] <- geneDataIn$normRatio[firstNonNAN]
	lastNonNAN <- which.max(is.na(geneDataIn$normRatio))
	geneDataIn$normRatio[-(1:lastNonNAN-1)] <- geneDataIn$normRatio[lastNonNAN-1] 
	#Scale the normRatio so that the median is 1 and adjust the count data
	geneDataIn$normRatio <- geneDataIn$normRatio/median(geneDataIn$normRatio)
	geneDataIn$normCounts <- round(geneDataIn$ReadCount/geneDataIn$normRatio)

	return(geneDataIn)
}

controlData$A <- add_norm_counts(controlData$A, 'MH_A')
controlData$B <- add_norm_counts(controlData$B, 'MH_B')
for (drugName in drugNames){
	drugData[[drugName]]$A <- add_norm_counts(drugData[[drugName]]$A, drugName)
	drugData[[drugName]]$B <- add_norm_counts(drugData[[drugName]]$B, drugName)
}


#Now we can start to import the data into edgeR
edgeRData <- data.frame(MH_A=controlData$A$normCounts, MH_B=controlData$B$normCounts)
for (drugName in drugNames){
	tmpVar <- paste(drugName, '_A', sep='')
	edgeRData[[tmpVar]] <- drugData[[drugName]]$A$normCounts
	tmpVar <- paste(drugName, '_B', sep='')
	edgeRData[[tmpVar]] <- drugData[[drugName]]$B$normCounts
}


rownames(edgeRData) <- rownames(controlData$A)
group <- c(rep("MH",2))
for (drugName in drugNames){
	group <- append(group, rep(drugName, 2))
}
group <- factor(group)
cds <- DGEList(edgeRData, group=group)

#Remove low counts as per edgeR vignette
cds <- cds[rowSums(1e+06 * cds$counts/expandAsMatrix(cds$samples$lib.size, dim(cds)) > 1) >= 3, ]
cds$samples$lib.size <- colSums(cds$counts)

#Now calculate normalization via TMM
cds <- calcNormFactors(cds)

plotMDS( cds )

cds <- estimateCommonDisp( cds, verbose=TRUE )

cds <- estimateTagwiseDisp( cds )

# # Open an Excel worksheet and write the data to it
# unlink(DEOutFile)
# wb <- loadWorkbook(DEOutFile, create=TRUE)

# for (drugName in drugNames){
# 	createSheet(wb, name=drugName)
# 	tmpET <- exactTest(cds, pair=c("MH", drugName))
# 	tmpData <- tmpET$table
# 	tmpData$PValue_adj <- p.adjust(tmpData$PValue, method='BH')
# 	tmpData$locus <- rownames(tmpData)
# 	writeWorksheet(wb, tmpData, sheet=drugName)
# }

# saveWorkbook(wb)
# rm(wb)

File: TnSeq/CalcDE.py
# Copyright 2013 Colm Ryan colm@colmryan.org
# License GPL v3 (http://www.gnu.org/licenses/gpl.txt) 

'''
Script to calculate real dVals
'''
from __future__ import division

import numpy as np
import pandas as pd
import rpy2.robjects as rBridge

#File names

#Control
controlName = 'MH'

#Drugs
drugNames = ['Vancomycin']

#First let's use the the all insertion site sheet to figure out the expected read count density
allGenesControlWB_A = pd.ExcelFile(controlName + '_Aggregate_A.xlsx')
allGenesControlWB_B = pd.ExcelFile(controlName + '_Aggregate_B.xlsx')

geneDataControl_A = allGenesControlWB_A.parse(allGenesControlWB_A.sheet_names[0], index_col=1, na_values=['']).drop('intergenic')
geneDataControl_B = allGenesControlWB_B.parse(allGenesControlWB_B.sheet_names[0], index_col=1, na_values=['']).drop('intergenic')


#Let's first look at read density (reads/length) to determine overal trend 

#Remove the top/bottom 10% spikes
qcuts = pd.qcut(geneDataControl_A['ReadCount'], [0, 0.1, 0.9, 1])
goodCts = geneDataControl_A[qcuts==qcuts.levels[1]]

goodCts['readDensity'] = goodCts['ReadCount']/goodCts['Length']

# #Push the data into R for the loess fit
rBridge.r.assign('positions', rBridge.IntVector(goodCts['Start'].values))
rBridge.r.assign('counts', rBridge.FloatVector(goodCts['readDensity'].values))
rBridge.r('loessFit <- loess(counts~positions, data.frame(positions=positions, counts=counts), control = loess.control(statistics = c("approximate"),trace.hat = c("approximate"), iterations=3))')
rBridge.r('smoothed <- predict(loessFit, positions)')

goodCts['smoothed'] = np.array(rBridge.r('smoothed'), dtype=np.float64)

# # # #Now bin the positions into 1kb pair regions weighted by the read counts to estimate the local read density
# xPts = np.arange(0,goodCts['Position'].max(),1000)
# densityEstimate = np.histogram(goodCts['Position'], bins=xPts, weights=goodCts['TotalCount'])[0]
# densityLoess = np.histogram(goodCts['Position'], bins=xPts, weights=goodCts['smoothed'])[0]


# #Fit this to a quadratic (should) probably update to some non-parametric approach like LOESS)
# fitDensity = np.poly1d(np.polyfit(xPts[:-1], densityEstimate,2))

# #Now we can load the gene data and calculate expected reads for each gene
# aggregateSheet = open_workbook(aggregateFile, on_demand=True).sheet_by_index(0)

# geneNameCol = 2
# lengthCol = 11
# startCol = 8
# ctCol = 7

# df = pd.DataFrame({	'start':[int(tmpValue) if tmpValue is not '' else 0 for tmpValue in aggregateSheet.col_values(startCol,1)],
# 					'length':[int(tmpValue) if tmpValue is not '' else 0 for tmpValue in aggregateSheet.col_values(lengthCol,1)],
# 					'readCounts':[int(tmpValue) if tmpValue is not '' else 0 for tmpValue in aggregateSheet.col_values(ctCol, 1)]},
# 					index=aggregateSheet.col_values(geneNameCol, 1) )


# geneMidPts = df['start']+df['length']/2

# df['expectedCts'] = df['length']/1000.0*fitDensity(geneMidPts)

# df['dVals'] = (df['readCounts']/df['expectedCts'])

# foldFactor = np.log2(df['dVals'])